File: js/ball.js
class Ball {
   constructor(x, y, r, speed, image, game) {
      this.x = x;
      this.y = y;
      this.r = r;
      this.dr = r + 2;
      this.s = speed;
      this.game = game;
      this.vx = speed * (Math.random() * 2 - 1);
      this.vy = -speed;
      this.image = image;
      this.countSteps = 0;
      const bound = $("main").getBoundingClientRect();
      this.offsetLeft = bound.left;
      this.offsetTop = bound.top;
      this.#setupNextCollision();
   }

   destroy() {
      // add particles
   }

   blockCollision({ x, y, w, h, isVisible }) {
      return (
         isVisible &&
         this.x + this.r >= x * w &&
         this.x - this.r <= x * w + w &&
         this.y + this.r >= y * h &&
         this.y - this.r <= y * h + h
      );
   }

   paddleCollision({ x, y, w, h }) {
      return (
         this.y + this.r > y &&
         this.y + this.r < y + h &&
         this.x + this.r > x - w / 2 &&
         this.x - this.r < x + w / 2
      );
   }

   xWallCollision() {
      return this.x + this.r > this.cvs.width || this.x - this.r < 0;
   }

   topWallCollision() {
      return this.y - this.r < 0;
   }

   #getMinPoint(A, B, startX, startY, endX, endY) {
      const a = new Point(startX, startY);
      const b = new Point(endX, startY);
      const c = new Point(endX, endY);
      const d = new Point(startX, endY);

      const intersects = [
         getIntersection(A, B, a, b, "top"),
         getIntersection(A, B, b, c, "right"),
         getIntersection(A, B, c, d, "bottom"),
         getIntersection(A, B, d, a, "left"),
      ].filter((e) => e);

      return intersects.sort((a, b) => a.offset - b.offset)[0];
   }

   #setupNextCollision() {
      const { cvs, paddle, blocks } = this.game;
      const { r } = this;

      const range = innerHeight;
      const angle = Math.atan2(-this.vx, this.vy) + Math.PI / 2; // ball angle in radians
      const ballTX = this.x + Math.cos(angle) * range;
      const ballTY = this.y + Math.sin(angle) * range;
      const A = new Point(this.x, this.y);
      const B = new Point(ballTX, ballTY);

      const intersectPoints = [];

      // boundary collision
      (() => {
         const minPoint = this.#getMinPoint(
            A,
            B,
            r,
            r,
            cvs.width - r,
            cvs.height - r
         );
         minPoint && intersectPoints.push(minPoint);
      })();

      // paddle collision
      (() => {
         const pa = new Point(0, paddle.y - this.r);
         const pb = new Point(cvs.width, paddle.y - this.r);

         const intersect = getIntersection(A, B, pa, pb, "top");
         intersect && intersectPoints.push(intersect);
      })();

      // block collision
      blocks.forEach((block) => {
         const { dx, dy, w, h, isVisible } = block;

         if (isVisible) {
            const minPoint = this.#getMinPoint(A, B, dx - r, dy - r, dx + w + r, dy + h + r);
            minPoint && intersectPoints.push({ ...minPoint, block }); 
         }
         // console.log(intersectPoints);
      }); 

      const mins = intersectPoints.sort((a, b) => a.offset - b.offset);
      let min = mins[1] ? mins[1] : mins[0]; 

      // if (Math.round(min.x) == Math.round(this.x) && Math.round(min.y) == Math.round(this.y)) {
      //    min = mins[1] || mins[0];
      // }

      console.log(mins);
      console.log(this.x, this.y); 
      console.log(min);

      const steps = Math.ceil(Math.abs((this.y - min.y) / this.vy) + 0.01); 

      console.log(steps);

      if (min.side === "left" || min.side === "right") this.vx *= -1;
      if (min.side === "top" || min.side === "bottom") this.vy *= -1;

      this.countSteps = steps;
      this.fixSteps = steps;
      this.tx = min.x;
      this.ty = min.y;
      this.ox = this.x;
      this.oy = this.y;

      return {
         steps,
         x: min.x,
         y: min.y,
         block: min.block ? min.block : null,
      };
   }

   update() {
      this.countSteps--;
      console.log(this.countSteps);  

      const p = new PointDraw(this.tx, this.ty);
      p.draw(ctx, "P", false, this.dr); 

      this.x += (this.tx - this.ox) * (1 / this.fixSteps);
      this.y += (this.ty - this.oy) * (1 / this.fixSteps);

      if (this.countSteps == 0) {
         this.x = this.tx;
         this.y = this.ty;
         console.log("target Complete");
         const { steps, x, y, side } = this.#setupNextCollision();
      }
   }

   draw(ctx) {
      ctx.drawImage(
         this.image,
         this.x - this.r,
         this.y - this.r,
         this.dr * 2,
         this.dr * 2
      );
   }
}

File: cpp/ball.cpp
#include "./ball.h"
#include "./random.h"
#include <algorithm>
#include <cmath>
#include <iostream>

void Ball::init(float _x, float _y, short _r, float _speed) {
   x = _x;
   y = _y;
   preX = _x;
   preY = _y;
   r = _r;
   speed = _speed;
   vx = rnd(-speed / 4, speed / 4);
   vy = -speed;
   this->updateX1X2Y1Y2();
}
void Ball::reset(float _x, float _y) {
   x = _x;
   y = _y;
   preX = _x;
   preY = _y;
   vx = rnd(-speed / 4, speed / 4);
   vy = -speed;
   this->updateX1X2Y1Y2();
}

void Ball::updateX1X2Y1Y2() {
   x1 = x - r;
   y1 = y - r;
   x2 = x1 + r * 2;
   y2 = y1 + r * 2;
}

void Ball::draw(DrawBallPtr drawBall) {
   drawBall(x, y, r);
}

void Ball::update() {
   preX = x;
   preY = y;
   x += vx;
   y += vy;
   this->updateX1X2Y1Y2();
}
void Ball::reverseY() {
   vy *= -1;
}
void Ball::reverseX() {
   vx *= -1;
}
void Ball::goLeft() {
   vx = abs(vx) * -1;
}
void Ball::goRight() {
   vx = abs(vy);
}
void Ball::goTop() {
   vy = abs(vy) * -1;
}
void Ball::goBottom() {
   vy = abs(vy);
}

short Ball::getCollisionSide(float X1, float Y1, float X2, float Y2, float offset) {
/* 
   Left: 1
   Right: 2
   Top: 3
   Bottom: 4
*/
   float left = x2 - X1;
   float right = X2 - x1;
   float top = y2 - Y1;
   float bottom = Y2 - y1;

   std::vector<float> numbers = {left, right, top, bottom};
   std::sort(numbers.begin(), numbers.end());

   float min = numbers[0];
   float second = numbers[1];
   float dx = x - preX;
   float dy = y - preY;

   if (abs(min - second) < offset) {
      if (min > offset / 2) {
         if (abs(dx) > abs(dy)) {
            return (left < right) ? 1 : 2; // Return 1 for left, 2 for right
         } else {
            return (top < bottom) ? 3 : 4; // Return 3 for top, 4 for bottom
         }
      }
   } else {
      if (min == left) {
         return 1; // Left
      } else if (min == right) {
         return 2; // Right
      } else if (min == top) {
         return 3; // Top
      } else if (min == bottom) {
         return 4; // Bottom
      }
   }

   return 0; // No collision
}

short Ball::checkPaddleCollision(Paddle *p) {
   if (p->x1 < x2 && p->x2 > x1 && p->y1 < y2 && p->y2 > y1) {
      return getCollisionSide(p->x1, p->y1, p->x2, p->y2, p->offset / speed);
   }
   return 0;
}

short Ball::checkBlockCollision(Block *b) {
   if (b->x1 < x2 && b->x2 > x1 && b->y1 < y2 && b->y2 > y1) {
      return getCollisionSide(b->x1, b->y1, b->x2, b->y2, b->offset / speed);
   }
   return 0;
}